pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+
+    def visiting_potential_leaf(self, node):
+        """
+        Visiting a potential leaf.
+
+        Return True if the node is not expanded.
+        """
+        if node.__class__ in nonpyomo_leaf_types or not node.is_potentially_variable():
+            return True, 0
+
+        if not node.is_expression_type():
+            return True, 0 if node.is_fixed() else 1
+
+        return False, None
+
+
+def _polynomial_degree(node):
+    """
+    Return the polynomial degree of the expression.
+
+    Args:
+        node: The root node of an expression tree.
+
+    Returns:
+        A non-negative integer that is the polynomial
+        degree if the expression is polynomial, or :const:`None` otherwise.
+    """
+    visitor = _PolyDegreeVisitor()
+    return visitor.dfs_postorder_stack(node)
+
+
+# =====================================================
+#  _expression_is_fixed
+# =====================================================
+
+class _IsFixedVisitor(ExpressionValueVisitor):
+    """
+    NOTE: This doesn't check if combiner logic is
+    all or any and short-circuit the test.  It's
+    not clear that that is an important optimization.
+    """
+
+    def visit(self, node, values):
+        """ Visit nodes that have been expanded """
+        return node._is_fixed(values)
+
+    def visiting_potential_leaf(self, node):
+        """
+        Visiting a potential leaf.
+
+        Return True if the node is not expanded.
+        """
+        if node.__class__ in nonpyomo_leaf_types or not node.is_potentially_variable():
+            return True, True
+
+        elif not node.is_expression_type():
+            return True, node.is_fixed()
+
+        return False, None
+
+
+def _expression_is_fixed(node):
+    """
+    Return the polynomial degree of the expression.
+
+    Args:
+        node: The root node of an expression tree.
+
+    Returns:
+        A non-negative integer that is the polynomial
+        degree if the expression is polynomial, or :const:`None` otherwise.
+    """
+    visitor = _IsFixedVisitor()
+    return visitor.dfs_postorder_stack(node)
+
+
+# =====================================================
+#  expression_to_string
+# =====================================================
+
+class _ToStringVisitor(ExpressionValueVisitor):
+
+    def __init__(self, verbose, smap, compute_values):
+        super(_ToStringVisitor, self).__init__()
+        self.verbose = verbose
+        self.smap = smap
+        self.compute_values = compute_values
+
+    def visit(self, node, values):
+        """ Visit nodes that have been expanded """
+        tmp = []
+        for i,val in enumerate(values):
+            arg = node._args_[i]
+
+            if arg is None:
+                tmp.append('Undefined')                 # TODO: coverage
+            elif arg.__class__ in native_numeric_types:
+                tmp.append(val)
+            elif arg.__class__ in nonpyomo_leaf_types:
+                tmp.append("'{0}'".format(val))
```

Comment:
Why use `.format()`?  It is measurably slower than `"'%s'" % (val,)`

Reply:
That's probably true.  I haven't worried about performance of string operations in expr_pyomo5.py, since this only impacts debugging info right now.